### PR TITLE
Add Averager load balancing and public endpoints

### PR DESCRIPTION
**Pull request by justheuristic**
* averager can now specify available network throughput at init
  * this parameter is used to dynamically split the vector
* averagers can now gather a small metadata message from their groupmates (passed through the leader)

### CODE REVIEW

File: tests/test_averaging.py [hunk 1]
```diff
@@ -112,10 +141,50 @@ def test_partitioning():
...
+
+@pytest.mark.forked
+def test_load_balancing():
+    def get_cost(vector_size, partitions, throughputs):
+        return max((vector_size - partitions[i] + (len(partitions) - 1) * partitions[i]) / max(throughputs[i], 1e-9)
+                   for i in range(len(partitions)))
+
+    def check_optimality(vector_size, throughputs, ref_partitions):
+        partitions = list(load_balance_peers(vector_size, throughputs))
+        assert get_cost(vector_size, partitions, throughputs) <= get_cost(vector_size, ref_partitions, throughputs)
```
Comment (mryab) [inline tests/test_averaging.py:162]: Extract to private functions outside the test

File: hivemind/client/averaging/matchmaking.py [hunk 2]
```diff
@@ -450,3 +478,5 @@ def compute_schema_hash(tensors: Sequence[torch.Tensor]) -> bytes:
                      for field_name, field_value in asdict(TensorDescriptor.from_tensor(tensor)).items()}
                     for tensor in tensors]
     return DHTID.generate(source=MSGPackSerializer.dumps(schema_dicts)).to_bytes()
+
+
```
Comment (mryab) [inline hivemind/client/averaging/matchmaking.py:482]: Redundant newline

File: hivemind/client/averaging/load_balancing.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    In butterfly all-reduce, each peer acts both as a "client" and as an "aggregator":
+    * a "client" splits his local vector into shards and sends each shard to one peer, then downloads the average
+    * an "aggregator" receives a certain part of vector components from all peers, aggregates and returns the average
+
+    Peer i network load as a "client" = vector_size * (1 - fraction_assigned_to_peer_i)
+    Peer i network load as an "aggregator" = vector_size * (group_size - 1) * fraction_assigned_to_peer_i
+    Peer i total communication = vector_size * [1 + (group_size - 2) * fraction_assigned_to_peer_i]
+    Total time = max_i (total_communication_for_peer_i / throughputs[i])
+
+    We find optimal vector fractions by minimizing the total time (using https://tinyurl.com/minimax-to-lp )
```
Comment (mryab) [inline hivemind/client/averaging/load_balancing.py:43]: We'd better give a direct link to the lecture notes

File: hivemind/client/averaging/load_balancing.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+    c = np.zeros(num_variables)
+    c[-1] = 1.0  # optimize w.r.t. ksi
+
+    # the constraints below are tuples (A, b) such that Ax <= b
+    nonnegative_weights = -np.eye(group_size, M=num_variables), np.zeros(group_size)
+    weights_sum_to_one = c[None, :] - 1.0, np.array([-1.0])
+    coeff_per_variable = (group_size - 2.0) / np.maximum(throughputs, eps)
+    coeff_matrix_minus_ksi = np.hstack([np.diag(coeff_per_variable), -np.ones((group_size, 1))])
+    ksi_is_maximum = coeff_matrix_minus_ksi[is_nonzero], -1.0 / throughputs[is_nonzero]
```
Comment (mryab) [inline hivemind/client/averaging/load_balancing.py:63]: ksi -> xi

File: hivemind/client/averaging/__init__.py [hunk 5]
```diff
@@ -215,19 +231,17 @@ async def _step(self, *, future: MPFuture, allow_retries: bool, timeout: Optiona
...
         """
         assert allreduce_group.return_deltas and allreduce_group.future.done()
         averaging_deltas = allreduce_group.future.result()
 
         with torch.no_grad(), self.get_tensors() as local_tensors:
             assert len(local_tensors) == len(self._averaged_tensors)
             for tensor, update in zip(local_tensors, averaging_deltas):
-                tensor.add_(update, alpha=self.averaging_alpha)
+                tensor.add_(update, alpha=self._averaging_alpha)
             return True
```
Comment (mryab) [inline hivemind/client/averaging/__init__.py:245]: Does this function still return True?

File: hivemind/client/averaging/load_balancing.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    force_max_weights = np.eye(group_size, M=num_variables), is_nonzero.astype(c.dtype)
+
+    A, b = list(map(np.concatenate, zip(nonnegative_weights, weights_sum_to_one, ksi_is_maximum, force_max_weights)))
+
+    solution = scipy.optimize.linprog(c, A_ub=A, b_ub=b)
+    if solution.success:
+        peer_fractions = solution.x[:group_size]
+        if np.max(peer_fractions) >= min_size / float(vector_size):
+            peer_fractions[peer_fractions < min_size / float(vector_size)] = 0.0
```
Comment (mryab) [inline hivemind/client/averaging/load_balancing.py:72]: Do the fractions still sum to 1 in that case?

File: hivemind/client/averaging/load_balancing.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    else:
+        logger.error(f"Failed to solve load-balancing for bandwidths {throughputs}.")
+        peer_fractions = np.ones(group_size) / group_size
+
+    return peer_fractions[np.argsort(permutation)]
+
+
+def hagenbach_bishoff(vector_size: int, scores: Sequence[float]) -> Sequence[int]:
```
Comment (mryab) [inline hivemind/client/averaging/load_balancing.py:80]: We should probably add a reference to https://github.com/crflynn/voting in the doctsting

**Review by mryab — CHANGES_REQUESTED** [review]

File: hivemind/client/averaging/load_balancing.py (hunk 6 above)
Comment (justheuristic) [reply hivemind/client/averaging/load_balancing.py:71]: renormalized on the next line

File: hivemind/client/averaging/load_balancing.py (hunk 7 above)
Comment (justheuristic) [reply hivemind/client/averaging/load_balancing.py:80]: done

**Review by mryab — APPROVED** [review]

**Pull request merged**